### ISSUE TITLE
Add option to ignore git repos

### ISSUE THREAD

**Issue by jonnykl** [nextcloud/desktop#4889]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you want to have the same feature implemented.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Feature description

I don't want to sync git repos. An option to ignore all git repos would solve this issue. There are multiple ways to implement this feature:

- ignore all directories containing a .git directory (if git repo ignore option is enabled)
- option to ignore parent directory to the existing ignore list entries -> this way you can add .git to the ignore list and enable that new option

**Comment by smihael** [comment]
I have a similar feature proposal.  Instead of just ignoring git repositories, nextcloud could sync the list of repositories and where in the file structure they should be located. A plugin for the client would then invoke git to clone/pull the contents of the repository. In the Web UI, one could be redirected to a web-interface associated with the git repository (if any).

I outlined a scenario, where such a feature would be beneficial at
https://softwarerecs.stackexchange.com/questions/85360/synchronisation-of-files-or-file-tree-structure-including-git-repositorieis

**Comment by bastian2001** [comment]
IMO the big problem with git repos is the many small files introduced by git. I think an even better approach would be to use `.gitignore` files and also just don't sync `.git` directories, rather than omitting the whole repo. If that's not desired, it would also be possible to make the exclusion list compatible with the gitignore syntax and just call the file somthing else, like `.ncignore`. I have a lot of problems with the sync exclusion list (files and directories being synced even though they are in the exclusion list, even after fully restarting the synchronisation), so that in my opinion something that is a more well-known standard like gitignore would be far superior and could also tackle individual folders instead of the whole synced directory with all subdirectories.

**Comment by Oliver-Habersetzer** [comment]
I downloaded the sourcecode and I'll try to integrate this.

My idea was also to call
```
git init
git remote add origin <remote_url>
```

That way a user can interact with the git repository as they're used to.

I just worry about branches: if you have unsaved changes on two PCs in different branches, we might get oscillating behavior that'd likely break the git repo. I'll have to test it a bunch.
- An alternative would be to create subfolders for each branch the nextcloud-desktop client has seen checked out - tho that'd obviously also inflate the file size of repos by a lot.
- Or unsaved changes are always uploaded to the server in a special folder for each branch/revision number. While you're still editting the files, they will stay there until you commit.
- Or it only uploads stashes/shelves of for uncommitted changes. it's more work for the user to get their current working state tho.

**Comment by mlasch** [comment]
An option like "Sync git repositories", which can be disabled, would be very helpful to exclude the whole Git workspace. Git uses its own mechanism to sync files over the network with other machines.

Working with Git, or a similar version control system, inside a synced directory reacts on Git operations like branch checkouts and would trigger updates on all other machines altering the working environment.

With such a "Sync git repositories" or a more general "Sync SVC repositories" option Git could be used independently inside a synced Nextcloud directory.

**Comment by joshtrichards** [comment]
Is there a reason the existing *Ignored Files* feature can't be used to accomplish this?

**Comment by smihael** [comment]
Yes, there is a reason. Last time I checked, there was nowhere to set files
excluded from the ignore pattern. I want to exclude the content of any git
repositories, *except for* the . git/config file.

**Comment by mlasch** [comment]
> Is there a reason the existing Ignored Files feature can't be used to accomplish this?

I don't think you can't accomplish this with just the _Ignored Files_. While it is possible to ignore the `.git` directory in a checked out repository, as far as I can tell, it is not possible to "ignore a whole directory which contains a .git". The reason is not only the `.git` should be ignored, but all files under version control.

This might go even further to still sync `.git/config` as mentioned by @smihael as well as "untracked files". But a basic "Don't sync git repositories" would be already a big step.

**Comment by Rello** [comment]
its possible with a workaround. please check here
https://github.com/nextcloud/desktop/issues/8659#issuecomment-3245965904

**Comment by Migsi** [comment]
> its possible with a workaround. please check here [#8659 (comment)](https://github.com/nextcloud/desktop/issues/8659#issuecomment-3245965904)

Becuase every repo comes with such an exclusion file as part of it's root, suuure 😏

I get that the nextcloud team doesn't want to introduce specific treatments for specific data structures. But what if we don't treat code repositories differently at all? The requested "ignore git repositories" setting could simply check any newly discovered directory for being a git repo and if it deretmines to have found one, drop a (ideally customizable) `.sync-exclude.lst` in there. Then the client can proceed as usual while leveraging existing ignore patterns.

Granted this will modify every cloned repo's contents, but as the exclusion list is never checked in, it should be pretty clear what happened. If a repo already contains such a file, simply go with that, never overwrite.

What do you think? Idea good enough to reopen the request?
